theano/tensor/basic.py

```diff
@@ -4126,9 +4126,14 @@ def make_node(self, x, shp):
         x = as_tensor_variable(x)
         shp_orig = shp
         shp = as_tensor_variable(shp, ndim=1)
-        if not shp.dtype.startswith('int'):
+        if not shp.dtype.startswith('int') and not(isinstance(shp,
+                                                   TensorConstant) and
+                                                   shp.data.size == 0):
```

Comment:
I think it would be more readable this way
```python
if (not shp.dtype.startswith('int') and
    not (isinstance(shp, TensorConstant) and shp.data.size == 0)):
```
And you could even factor out the negation
```python
if not (shp.dtype.startswith('int') or
        (isinstance(shp, TensorConstant) and shp.data.size == 0)):
```